[Chai Yew Meng] iP

yewmeng28 opened the pull request:
### Necro

> “Do or do not. There is no try.” – Master Yoda, _The Return of the Jedi_

Necro removes the hassle of needlessly remembering your menial tasks. It's,

- text-based
- easy to learn
- guided by help commands
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from here. (add hyperlink)
2. **double-click** it.
3. add your tasks.
4. let it manage your tasks for you ;)

And it is **FREE!!!** WOOOHOOOOO!!!!!

Features:

- [X] Managing tasks
- [X] Managing deadlines
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

**SIKE!!! THAT'S THE WRONG METHOD!!! OHHHHHHHHHHHHHHH**

Acknowledgements
- JSON Object to String converter: methods adapted from user huizhuansam
- JavaFX: Adapted from the SE-EDU Java Guides
- Necro icon: (Darth Nihilus Mask PNG) edited from https://www.pngkey.com/maxpic/u2w7q8w7u2a9e6r5/
- User icon: (Darth Revan Mask PNG) edited from https://custom-cursor.com/en/collection/star-wars/sw-revan-lightsaber

Commit:
Add Gradle support

yewmeng28 committed:
Necro chatbot

yewmeng28 committed:
Necro list

yewmeng28 committed:
Necro Mark as Done

yewmeng28 committed:
Necro ToDo, Event, Deadline

yewmeng28 committed:
Update test files

yewmeng28 committed:
Add handling of errors

yewmeng28 committed:
Add Delete functionality, change array to ArrayList<Task>

yewmeng28 committed:
Add classes for the Save function

yewmeng28 committed:
Add classes for Save function

yewmeng28 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-{Level-7}

yewmeng28 committed:
Refactor old code and add classes for the Save function

yewmeng28 committed:
Refactor more code and add working Save functionality

yewmeng28 committed:
Add Date and Time formatting for Deadlines

tanjiaxian99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,266 @@
+    private static final String tab = "      ";
+    private static final String line = "------------------------------------------------------------";
+
+    public static void textBox(String... messages) {
```
Maybe consolidate all the UI stuff into another class?

tanjiaxian99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,266 @@
+        try {
+            tasks = new TaskList(storage.load());
+        } catch (FileNotFoundException e) {
+            tasks = new TaskList(new ArrayList<>());
```
Will the user be informed if the file cannot be found?

tanjiaxian99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,266 @@
+                    }
+                    System.out.println(tab + line);
+                }
+            } else if (input.equals("bye")) {
```
Consider having a Parser class that takes care of the different inputs?

tanjiaxian99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,266 @@
+                    // Check if there are any spaces first. if there are no spaces, means that it is either an invalid input or no such task
+
+                    // Level-5
+                    try {
```
Is there a way to abstract away all the try-catch clauses?

tanjiaxian99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,266 @@
+
+    }
+
+    public static class Storage {
```
Maybe make the Storage class a proper class instead of keeping it nested?

tanjiaxian99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,266 @@
+            this.filePath = filePath;
+        }
+
+        @SuppressWarnings("unchecked")
```
Perhaps shift the @SupressWarnings to the exact line which you would like to suppress?

tanjiaxian99 reviewed (commented):
I think more abstraction can be done to make the codebase cleaner. Code quality can definitely be improved.

czhi-bin commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,266 @@
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+import java.io.*;
```
Perhaps a single class import for this might help to make it clearer as to which function you are using from the java.io package

czhi-bin commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,266 @@
+                + "| | \\ \\| | |    __/  |  <_   |  /    |   |_|   |\n"
+                + "|_|  \\___|  \\_____|  \\____|  |__|     \\  ___  / \n";
+        System.out.println("Hello from\n" + necroLogo);
+        String tab = "      ";
+        String line = "------------------------------------------------------------";
```
I think these 2 variables are already declared in the class fields? Maybe you can just reuse those variables?

czhi-bin commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,266 @@
+        System.out.println(tab + line);
+    }
+
+    public static void main(String[] args) throws NoSuchTaskException, InvalidInputException, FileNotFoundException {
```
Perhaps adding a JavaDoc could help people to understand your code?

czhi-bin reviewed (commented):
I think there is not much coding standard violation, but perhaps you could add in some JavaDoc for public classes and method to help people understand better what does your code do?

yewmeng28 committed:
Improve OOP

yewmeng28 committed:
Make the code follow more OOP principles

yewmeng28 committed:
Add JUnit Tests and outputs to a Jar file

yewmeng28 committed:
Add JavaDocs to most methods

yewmeng28 committed:
Conform to coding standard for the switch statement

yewmeng28 committed:
Add Find function

yewmeng28 committed:
Revert "Add Find function"

This reverts commit c773251131663d50f4e0cc0513f2f65288a93fa8.

yewmeng28 committed:
Revert "Revert "Add Find function""

This reverts commit 79777497f8e925e7b0125d424a476f4da3041076.

muhammad-faruq commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,22 @@
+import duke.util.Ui;
+import duke.util.TaskList;
+
+public abstract class Command {
```
Best to include javadocs for your class to explain why you have a command class and what is the purpose of this command class

```suggestion
/**
 * Command abstract class as a base for other commands class.
 * These command classes is where the logic of the command and its output will be written.
 * Every command class must have an execute method.
 */
public abstract class Command {
```

muhammad-faruq commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,22 @@
+     */
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws InvalidInputException;
+
+    public abstract boolean isExit();
```
isExit is a rather odd boolean name. Perhaps use something more natural such as isRunning or willExit

muhammad-faruq commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,55 @@
+            try {
+                if (strings.length < 2) {
+                    ui.textBox("You did not specify which task to complete.");
+                } else {
```
In this case, perhaps it would be better to not pass the command word into the `String input` .
This is because you would have already checked the command word in the parser hence there is no need to pass it in again into the command object. In this case there will be no need to call split again hence making this a bit more efficient.

muhammad-faruq commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,51 @@
+     * @throws InvalidInputException When the input is deemed invalid.
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws InvalidInputException {
```
Perhaps you can make this part more user friendly by allowing users to search for the task reagardless of capitalization. 

So in that case if i search "apple", the program should still return "Apple Eating Contest" or "Work at Apple"

muhammad-faruq commented on src/main/java/duke/exception/InvalidInputException.java:
```diff
@@ -0,0 +1,14 @@
+package duke.exception;
+
+public class InvalidInputException extends Exception {
```
Perhaps you can explain more on what causes this error. Where will the invalid input be from?

muhammad-faruq commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,84 @@
+
+
+    @Override
+    public String toString() throws DateTimeException {
```
By the looks of it, if the user inputs an invalid date, this will cause this error to be thrown causing the program to stop. Perhaps catch the error and return something more useful to the user such as how to format the date properly to make the program work.

muhammad-faruq commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,96 @@
+            }
+            break;
+        case "deadline":
+            try {
```
As explained in deadline task, the error on improper date time formatting will cause the program to stop. Catch this error to return something useful for the user to work with.

muhammad-faruq reviewed (commented):
Code quality review.
Some improvements can be done to avoid errors and frustrations on the user side.

muhammad-faruq commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
```
Javadocs needed for this method

muhammad-faruq commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @param index The index of the Task we want to get.
+     * @return The Task with the specified index.
+     */
```
vague naming convention. Perhaps try to use things such as getTask or something that explains what you are trying to get

muhammad-faruq replied:
okay I realize its actually okay cos this is tasklist

muhammad-faruq commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @Override
+    public boolean isExit() {
```
javadocs for this method

muhammad-faruq reviewed (commented):
Coding standards violation review. 

Overall not much errors. Some clarification can be done to further explain the purpose of the class and the purpose of some methods that may seem trivial

yewmeng28 committed:
Merge branch 'branch-A-CodingStandard'

yewmeng28 committed:
Add Gradle support

yewmeng28 committed:
Add Checkstyle to the Grade build

yewmeng28 committed:
Add GUI

yewmeng28 committed:
Check the branching

yewmeng28 committed:
Update Level 10 progress

yewmeng28 committed:
Add a working GUI

yewmeng28 committed:
Update README.md

yewmeng28 committed:
Update README.md

Test if the progress chart updates

yewmeng28 committed:
Set theme jekyll-theme-merlot

yewmeng28 committed:
Add User Guide and improve GUI

yewmeng28 committed:
Merge branch 'patch-1' into branch-Level-10

yewmeng28 committed:
Merge branch 'master' into branch-Level-10

# Conflicts:
#	docs/README.md

yewmeng28 committed:
Improve GUI

yewmeng28 committed:
Add Assertions

yewmeng28 committed:
Add minor Code Quality improvements

yewmeng28 committed:
Merge branch 'branch-A-Assertions' into branch-A-CodeQuality

yewmeng28 committed:
Merge pull request #3 from NecroWolf28/branch-A-Assertions

Add Assertions

yewmeng28 committed:
Merge pull request #4 from NecroWolf28/branch-A-CodeQuality

Add minor improvements to code

yewmeng28 committed:
Revert "Improve GUI"

This reverts commit 79425bd6abe657894c14a0c01ed968a9ea206fd4.

yewmeng28 committed:
Revert "Revert "Improve GUI""

This reverts commit 46146106febe72f8c7bea64c6965c9b6511f3a1e.

yewmeng28 committed:
Merge remote-tracking branch 'origin/master'

yewmeng28 committed:
Add the C-Help function. Also add another full commit message.

yewmeng28 committed:
Delete unused files. Also include the full commit message that the
previous commit did not include.

Previously there were no Help commands, but this commit will enable
the user to seek help from Necro.

This will improve user experience as he can be guided on how to use the
application to its fullest capabilities.

yewmeng28 committed:
Update README.md

Previously this commit did not include the updated description.

This commit will update the description, which is still to be updated.

yewmeng28 committed:
Update README.md

Add acknowledgements

yewmeng28 committed:
Update User Guide

Updates the User Guide.

The previous User Guide was not complete. This commit will include all of its features, including the C-Help extension.

yewmeng28 committed:
Add acknowledgements for code snippets, and edit some Help related code.

yewmeng28 committed:
Merge branch 'master' of https://github.com/NecroWolf28/ip

yewmeng28 committed:
Add the Ui.png file.

yewmeng28 committed:
Fix some Ui related bugs. (For example, saying there are more tasks when
the tasks list is of size 0)
